types/src/transaction/authenticator.rs

```diff
@@ -17,19 +17,179 @@ use rand::{rngs::OsRng, Rng};
 use serde::{Deserialize, Serialize};
 use std::{convert::TryFrom, fmt, str::FromStr};
 
-/// A `TransactionAuthenticator` is an an abstraction of a signature scheme. It must know:
+/// Each transaction submitted to the Diem blockchain contains a `TransactionAuthenticator`. During
+/// transaction execution, the executor will check if every `AccountAuthenticator`'s signature on
+/// the transaction hash is well-formed and whether the sha3 hash of the
+/// `AccountAuthenticator`'s `AuthenticationKeyPreimage` matches the `AuthenticationKey` stored
+/// under the participating signer's account address.
+
+#[derive(Clone, Debug, Eq, PartialEq, Hash, Serialize, Deserialize)]
+pub enum TransactionAuthenticator {
```

Comment:
I don't think we can change this `TransactionAuthenticator ` as this is part of the `SignedTransaction`. That means we can only append variants to existing `TransactionAuthenticator` enum.

Reply:
cc @davidiw

Reply:
Right. I just realized the backward compatibility issue here. I will modify this PR so that we keep the variants TransactionAuthenticator originally had and add single agent and multi agent variants.